src/plots/plots.js

```diff
@@ -609,7 +592,61 @@ function relinkPrivateKeys(toLayout, fromLayout) {
     }
 }
 
-plots.supplyDataDefaults = function(traceIn, i, layout) {
+plots.supplyDataDefaults = function(dataIn, dataOut, layout, modules) {
+    var cnt = 0;
+
+    // push to array if item isn't already in array
+    function fill(arr, item) {
+        if(item && arr.indexOf(item) === -1) return arr.push(item);
+    }
+
+    // detect plot type
+    function detect(trace) {
+        if(plots.traceIs(trace, 'cartesian')) layout._hasCartesian = true;
+        else if(plots.traceIs(trace, 'gl3d')) layout._hasGL3D = true;
+        else if(plots.traceIs(trace, 'geo')) layout._hasGeo = true;
+        else if(plots.traceIs(trace, 'pie')) layout._hasPie = true;
+        else if(plots.traceIs(trace, 'gl2d')) layout._hasGL2D = true;
+        else if(plots.traceIs(trace, 'ternary')) layout._hasTernary = true;
+        else if('r' in trace) layout._hasPolar = true;
+    }
+
+    for(var i = 0; i < dataIn.length; i++) {
+        var trace = dataIn[i];
+
+        var fullTrace = plots.supplyTraceDefaults(trace, cnt, layout);
+
+        // keep track of pre-transform _input
+        var traceIn = fullTrace._input;
+
+        if(fullTrace.transforms && fullTrace.transforms.length) {
+            var expandedTraces = applyTransforms(fullTrace, dataOut, layout);
+
+            for(var j = 0; j < expandedTraces.length; j++) {
+                var expandedTrace = expandedTraces[j];
+                var fullExpandedTrace = plots.supplyTraceDefaults(expandedTrace, cnt, layout);
+
+                // copy refs
+                fullExpandedTrace._input = traceIn;
+                fullExpandedTrace._fullTransforms = fullTrace.transforms;
+                fullExpandedTrace._index = i;
+
+                dataOut.push(fullExpandedTrace);
+                fill(modules, fullExpandedTrace._module);
+                detect(fullExpandedTrace);
+                cnt++;
+            }
+        }
+        else {
+            dataOut.push(fullTrace);
+            fill(modules, fullTrace._module);
+            detect(trace);
```

Comment:
in fact, this untransformed case needs `i` linked in as well. Might as well make eg:
```javascript
function addFinalTrace(finalTrace) {
    finalTrace._input = trace;
    finalTrace._inputIndex = i;
    finalTrace._expandedIndex = cnt;
    dataOut.push(finalTrace);
    fill(modules, finalTrace);
    detect(finalTrace);
    cnt++;
}
```
since all of that needs to happen identically whichever way you added a trace to `_fullData`.